csharp/src/Ice/ObjectAdapter.cs

```diff
@@ -1193,6 +1068,132 @@ private void UpdateLocatorRegistry(LocatorInfo? locatorInfo, IObjectPrx? proxy)
+                Console.Out.WriteLine($"{Name} ready");
+            }
+
+            lock (_mutex)
+            {
+                Debug.Assert(_state == State.Activating);
+                foreach (IncomingConnectionFactory factory in _incomingConnectionFactories)
+                {
+                    factory.Activate();
+                }
+                _state = State.Active;
+            }
+        }
+
+        private async Task PerformDeactivateAsync()
+        {
+            // Wait for activation to complete. This is necessary avoid out of order locator updates.
+            if (_activateTask != null)
+            {
+                await _activateTask.ConfigureAwait(false);
+            }
+
+            lock (_mutex)
+            {
+                Debug.Assert(_state <= State.Deactivating);
+                _state = State.Deactivating;
+            }
+
+            // Note: the router/locator infos and incoming connection factory list are immutable at this point.
+            try
+            {
+                if (_routerInfo != null)
+                {
+                    // Remove entry from the router manager.
+                    Communicator.EraseRouterInfo(_routerInfo.Router);
+
+                    // Clear this object adapter with the router.
+                    _routerInfo.Adapter = null;
+                }
+
+                await UpdateLocatorRegistryAsync(_locatorInfo, null);
+            }
+            catch
+            {
+                // We can't throw exceptions in deactivate so we ignore failures to update the locator registry.
+            }
+
+            lock (_mutex)
+            {
+                if (_directCount > 0)
+                {
+                    Debug.Assert(_directCountSemaphore == null);
+                    _directCountSemaphore = new SemaphoreSlim(_directCount);
+                }
+            }
+
+            if (_directCountSemaphore != null)
+            {
+                await _directCountSemaphore.WaitAsync().ConfigureAwait(false);
+            }
+
+            foreach (IncomingConnectionFactory factory in _incomingConnectionFactories)
+            {
+                _ = factory.DestroyAsync();
+            }
+
+            Communicator.OutgoingConnectionFactory().RemoveAdapter(this);
+
+            lock (_mutex)
+            {
+                Debug.Assert(_state == State.Deactivating);
+                _state = State.Deactivated;
+            }
+        }
+
+        private async Task PerformDisposeAsync()
+        {
+            await DeactivateAsync();
+
+            lock (_mutex)
+            {
+                // Only a single thread is allowed to destroy the object adapter. Other threads wait for the
+                // destruction to complete.
+                Debug.Assert(_state < State.Destroying);
+                _state = State.Destroying;
+            }
+
+            Communicator.RemoveObjectAdapter(this);
+
+            _directCountSemaphore?.Dispose();
```

Comment:
I would move this call after `_directCountSemaphore.WaitAsync()` to make sure `WaitAsync` returned before we call Dispose.